pkg/build/api/validation/validation.go

```diff
@@ -151,50 +161,27 @@ func hasProxy(source *buildapi.GitBuildSource) bool {
 
 func validateSource(input *buildapi.BuildSource) fielderrors.ValidationErrorList {
 	allErrs := fielderrors.ValidationErrorList{}
-	switch input.Type {
-	case buildapi.BuildSourceGit:
-		if input.Git == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("git"))
-		} else {
-			allErrs = append(allErrs, validateGitSource(input.Git).Prefix("git")...)
-		}
-		if input.Dockerfile != nil {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
-		}
+
+	if input.Git != nil {
+		allErrs = append(allErrs, validateGitSource(input.Git).Prefix("git")...)
 		if input.Binary != nil {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("binary", "", "may not be set when type is Git"))
-		}
-	case buildapi.BuildSourceBinary:
-		if input.Binary == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("binary"))
-		} else {
-			allErrs = append(allErrs, validateBinarySource(input.Binary).Prefix("binary")...)
-		}
-		if input.Dockerfile != nil {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("binary", "", "may not be set when git also set"))
 		}
+	}
+	if input.Binary != nil {
+		allErrs = append(allErrs, validateBinarySource(input.Binary).Prefix("binary")...)
 		if input.Git != nil {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("git", "", "may not be set when type is Binary"))
-		}
-	case buildapi.BuildSourceDockerfile:
-		if input.Dockerfile == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("dockerfile"))
-		} else {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("git", "", "may not be set when binary is also set"))
```

Comment:
You have some dead code there.

Suggestion: test earlier, in the function's top level, for the case where `input.Git != nil && input.Binary != nil && !isCustomStrategy` and return `fielderrors.NewFieldInvalid("", "", "git and binary sources may not be used together")`.

Note the first arguments to `NewFieldInvalid` are empty strings, so that the final message will refer to the `buildconfig.spec.source` field, instead of `buildconfig.spec.source.git` or `buildconfig.spec.source.binary`.